File: K_player/ml_zipf.py
#!/usr/bin/python
import numpy as np

def ml_zipf(xx):
	x = np.array(xx)
	
	mi = min(x)
	ma = max(x)
	
	l = .01
	u = 10.
	
	s = 10.
	
	f = []
	for i in range(mi,ma+1):
		f.append(sum(x==i)/len(x))
	
	suf = sum(f)
	su = sum(np.multiply(f,np.log(range(mi,ma+1))))
	
	error = 1000
	
	while error > 1e-5:
		ss = np.linspace(l,u,100)
		Hs = []
		for s in ss:
			Hs.append(sum(1/(np.power(range(mi,ma+1),s))))
		
		La = -np.multiply(ss,su) - suf*np.log(Hs)
			
		idx = np.argmin(La)
		l = max(l,ss[max(idx-1,0)])
		u = min(u,ss[min(idx+1,99)])
		error = u - l
	
	s = (l+u)/2
	
	return s
	

def ml_clauset(xx):
	x = np.array(xx)
	
	mi = min(x)
	ma = max(x)
	n = len(x)
	
	l = .01
	u = 10.
	s = 10.
	
	sum_data = sum(np.log(x))
	
	error = 1000
	
	while error > 1e-5:
		ss = np.linspace(l,u,100)
		zeta = []
		for si in ss:
			zeta.append(sum(np.power(range(mi,ma+1),-si)))
		
		L = -n*np.log(zeta) - np.multiply(ss,sum_data)
		
		idx = np.argmax(L)
		l = max(l,ss[max(idx-1,0)])
		u = min(u,ss[min(idx+1,99)])
		error = u - l
	
	s = (l+u)/2
	
	return s

File: 18c-k-player/ml_zipf.jl
# Computes the s coefficient of the Zipf's law with max likelihood for the data set x.

function ml_zipf(x::Array{Float64,1})
	mi = minimum(x)
	ma = maximum(x)
	
	l = .01
	u = 10.
	
	s = 10.
	
	f = Array{Float64,1}()
	for i in mi:ma
		push!(f,sum(x.==i)/length(x))
	end
	suf = sum(f)
	su = sum(f.*log(mi:ma))
	
	error = 1000
	
	while error > 1e-5
		ss = linspace(l,u,100)
		Hs = Array{Float64,1}()
		for s in ss
			push!(Hs,sum(1 ./((mi:ma).^s)))
		end
		La = -ss*su - suf*log(Hs)
		
		id = indmax(La)
		l = max(l,ss[max(id-1,1)])
		u = min(u,ss[min(id+1,100)])
		error = u - l
	end
	
	s = (l+u)/2

	return s
end

function ml_clauset(x::Array{Float64,1})
	mi = minimum(x)
	ma = maximum(x)
	n = length(x)
	
	l = .01
	u = 10.
	s = 10.
	
	sum_data = sum(log(x))
	
	error = 1000
	
	while error > 1e-5
		ss = linspace(l,u,100)
		zeta = Array{Float64,1}()
		for si in ss
			push!(zeta,sum((mi:ma).^(-si)))
		end
		
		L = -n*log(zeta) - ss.*sum_data
		
		id = indmax(L)
		l = max(l,ss[max(id-1,1)])
		u = min(u,ss[min(id+1,100)])
		error = u - l
	end
	
	s = (l+u)/2
	
	return s
end


